mqtt_as/interfaces/__init__.py

```diff
@@ -0,0 +1,92 @@
+from uerrno import EINPROGRESS, ETIMEDOUT
+import usocket
+import uasyncio as asyncio
+
+
+async def _g():
+    pass
+
+
+_type_coro = type(_g())
+
+
+# If a callback is passed, run it and return.
+# If a coro is passed initiate it and return.
+# coros are passed by name i.e. not using function call syntax.
+def launch(func, tup_args):
+    res = func(*tup_args)
+    if isinstance(res, _type_coro):
+        res = asyncio.create_task(res)
+    return res
```

Comment:
While I get while you would want functions but I deliberately chose to allow only async functions.
If you have a longer task and and you launch it with create task and it's duration is e.g. 10 secs you have a state change the next callback will run in parallel to the already launched tasks and anything can happen.
If you only allow async functions then you can at least try to cancel them.

Reply:
Sorry but I think you somehow mixed callbacks and coroutines in this answer and now it doesn't really make much sense but I think I get your point.
Callbacks would be the better option for wifi state changes because they have to be executed quickly without any delays in it.
Coroutines can cause the problem of multiple coroutines running in parallel if the state changes quickly, which is quite possible. Cancelling those is not implemented in this simple interface and the user can't cancel them either because the tasks are not bound anywhere.

Personally I'd go with callbacks only and not allow coroutines but I have to think about backwards compatibility and the original mqtt_as has a wifi_coro, a coroutine. So I have to allow it in order to not break existing implementations.

We have the same problem with the connected_coro in mqtt_as, which isn't a callback either. The only reason neither coroutines cause trouble in mqtt_as is that there's a 5 seconds delay in the wlan connect to ensure wlan stability. Most coroutines should have finished after 5 seconds. But if we use a different interface or reduce the wlan reconnects, this will become a problem. 
But again, backwards compatibility.. can't change that. Have to keep it that way. However, the wifi connected callback is now called directly after the connection is created and not after the broker was connected. This might result in more wifi callbacks as the wifi could get disconnected quickly if the connection is bad but I think it's an acceptable compromise and not many people will do much in the wifi connected callback because in mqtt_as this doesn't really mean anything because wifi will get disconnected again if the broker connection fails. (which is generally something I'd like to change in a more advanced interface because if your broker vanishes e.g. because you changed its ip adress or switched devices, your micropython device becomes inaccessible because you only have 5 seconds to establich a webrepl connection and interrupt the reconnect process..)


This is the reason why in my project, I subclassed the mqtt client and only have one connected coroutine that launches all my custom coroutines and also cancels all those tasks if the wifi gets disconnected again before it could finish (the coroutines just subscribe to the topics again, which would get paused if the client gets disconnected, so those would easily stack up).

Reply:
> Sorry but I think you somehow mixed callbacks and coroutines in this answer and now it doesn't really make much sense but I think I get your point.

Is a coro that is called on connect not a callback? I am not sure about the correct terminology.


> Personally I'd go with callbacks only and not allow coroutines but I have to think about backwards compatibility and the original mqtt_as has a wifi_coro, a coroutine. So I have to allow it in order to not break existing implementations.

If the has a long running sync callback things still can block, that's why I think a coro is better because it's immediately clear they have to be non blocking. And if it's all coros they have to be canceled by the base class.

Reply:
Callbacks in python (afaik) are always synchronous functions. Even the asyncio module in Python uses callbacks on certain Task events and those are all synchronous functions. Therefore my distinction between a callback=synchronous function and a coroutine=asynchronous function.

> If the has a long running sync callback things still can block, that's why I think a coro is better because it's immediately clear they have to be non blocking. And if it's all coros they have to be canceled by the base class.

A long sync function will block no matter where it is implemented. If someone writes blocking code, he will do so no matter if the context is a callback or a coroutine. 
If someone assumes he can write blocking code while using uasyncio, he doesn't undertand the environment he's programming in. 
That said, callbacks are always supposed to be short whereas when using coroutines, people might write long code, which is not blocking but can take several seconds to finish.
Of course, not a problem if that task can be cancelled, but that would require the user to be aware of it and make their code handle cancellation correctly, which could be problematic for people new to python/asyncio and therefore it's best to stay away from that. Also it would make the base class more complex, which I'd like to avoid, not many will need that feature. I myself only log the connection state to the console in my project. A very short callback.

(As mentioned elsewhere my connected coro is a lot longer and more complex and actually cancels all tasks if the connection state changes. But that isn't part of the hardware interface but mqtt_as..)